### PR TITLE
[native compiler] Remove `Lconstruct` from lambda code.

### PR DESCRIPTION
**Pull request by maximedenes**
This is a step towards unifying the higher level ILs of the native and
bytecode compilers.

See https://github.com/coq/coq/projects/17

### CODE REVIEW

**Comment by ppedrot** [comment]
How does this play with inductive cumulativity? Isn't systematically eta-expanding a constructor forbidding it to take place on parameters somehow?

**Comment by SkySkimmer** [comment]
Rather we should be systematically eta expanding to get it to work, cf https://github.com/coq/coq/issues/7015

**Comment by maximedenes** [comment]
> this

What do you mean? This patch doesn't change how/when we eta expand constructors, does it?

**Comment by ppedrot** [comment]
@maximedenes Sorry, I misread your patch. Indeed the generated code is the same, because `Gconstruct` definitions are always η-expanded.

File: kernel/nativecode.ml [hunk 1]
```diff
@@ -1280,9 +1280,6 @@ let ml_of_instance instance u =
   | Lmakeblock (prefix,(cn,_u),_,args) ->
      let args = Array.map (ml_of_lam env l) args in
      MLconstruct(prefix,cn,args)
-  | Lconstruct (prefix, (cn,u)) ->
-     let uargs = ml_of_instance env.env_univ u in
-      mkMLapp (MLglobal (Gconstruct (prefix, cn))) uargs
```
Comment (ppedrot) [inline kernel/nativecode.ml]: It looks like this is the only place where we generate a `Gconstruct` node, which means it is now dead code. You should probably remove it as well.

Comment (maximedenes) [reply kernel/nativecode.ml]: Oh, good point, let me do it as part of this PR, since it will simplify some further patches.

Comment (maximedenes) [reply kernel/nativecode.ml]: Done.

**Comment by maximedenes** [comment]
> This patch doesn't change how/when we eta expand constructors, does it?

In fact, this was a bit inaccurate: we now eta-expand on the fly, whereas we used to precompile eta-expanded versions. The VM was already doing the latter. I don't think it matters a lot, since performance-critical examples more often involve fully applied constructors.

**Review by ppedrot — APPROVED** [review]

**Pull request merged**